File: DSA/maximum-width-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * Time O(n) | Space O(n) | The time complexity is not O(n). 
 * We are assuming the array as a dequeue. 
 * the same code converted to python where we use dequeue instead of array has O(n) Time complexity.
 * https://leetcode.com/problems/maximum-width-of-binary-tree/
 * LevelOrder Traversal | BFS
 * @param {TreeNode} root
 * @return {number}
 */
var widthOfBinaryTree = function(root) {

    const trimNulls = (arr) => {
        
        while(arr.length && arr[arr.length - 1] === null) {
            arr.pop();
        }

        arr.reverse();

        while(arr.length && arr[arr.length - 1] === null) {
            arr.pop();
        }
        
        arr.reverse();

        return arr;
    }


    const bfs = (root) => {

        let maxWidth = 0;
        const q = [root];

        while(q.length) {
            trimNulls(q);
            
            let size = q.length;

            maxWidth = Math.max(maxWidth, size);

            while(size) {
                const node = q.shift();

                if(!node) {
                    q.push(null);
                    q.push(null);
                    size--;
                    continue;
                }

                if(node.left) q.push(node.left);
                if(!node.left) q.push(null);

                if(node.right) q.push(node.right);
                if(!node.right) q.push(null);

                size--;
            }

        }

        return maxWidth;
    }

    return bfs(root);
};

File: DSA/maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Time O(N) | Space O(N)

from typing import Optional
from collections import deque

class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        
        max_width = 0
        q = deque([(root, 0)])  # Store both node and its position index

        while q:
            level_size = len(q)
            start_idx = q[0][1]  # Get the index of the first node in the level
            end_idx = q[-1][1]  # Get the index of the last node in the level
            max_width = max(max_width, end_idx - start_idx + 1)  # Calculate the width of the current level

            for _ in range(level_size):
                node, idx = q.popleft()
                if node.left:
                    q.append((node.left, 2 * idx))  # Calculate the index for left child
                if node.right:
                    q.append((node.right, 2 * idx + 1))  # Calculate the index for right child

        return max_width
